### ISSUE TITLE
Not compatible with  .NET 8's dependency injection KeyedService

### ISSUE THREAD

**Issue by EashShow** [dotnetcore/CAP#1436]
升级到Net8的时候，报错。
这个类ConsumerServiceSelector，
这个方法FindConsumersFromInterfaceTypes
这个地方需要兼容一下Net8的KeyedService：foreach (var service in serviceCollection.Where(o => o.ImplementationType != null || o.ImplementationFactory != null))
直接获取实现类型，会报错的，需要判断一下是不是KeyedService

**Comment by yang-xiaodong** [comment]
Thanks for your feedback, would you want to submit a PR to fix it ?

**Comment by yang-xiaodong** [comment]
Sorry, I wasn't aware that we need to upgrade to .NET 8 as a whole. Ensuring compatibility with .NET 6 and .NET 7 is essential, so we must carefully consider adopting a compatible approach.

**Comment by EashShow** [comment]
Net8是长期支持版本，增加了很多好的特性。行业内已经有不少升级到Net8了。
这个兼容性升级优化能快些的话，就更好了。

**Comment by yang-xiaodong** [comment]
@EashShow 

Your PR can achieve compatibility with different frameworks using compatibility directives, and it only requires adjustments to the following two files

**DotNetCore.CAP.csproj**

```
<Project Sdk="Microsoft.NET.Sdk">
  
  <PropertyGroup>
    <TargetFrameworks>net6.0;net8.0</TargetFrameworks>
	<Nullable>enable</Nullable>
  </PropertyGroup>
  
  <PropertyGroup>
	<GenerateDocumentationFile>true</GenerateDocumentationFile>
    <NoWarn>1701;1702;1705;CS1591</NoWarn>
    <PackageReadmeFile>README.md</PackageReadmeFile>
  </PropertyGroup>
  
  <ItemGroup>
    <None Include="..\..\README.md">
      <Pack>True</Pack>
      <PackagePath>\</PackagePath>
    </None>
  </ItemGroup>
  
  <ItemGroup Condition=" '$(TargetFramework)' != 'net8.0' ">
    <PackageReference Include="Microsoft.Extensions.Hosting.Abstractions" Version="6.0.0" />
    <PackageReference Include="Microsoft.Extensions.Logging.Abstractions" Version="6.0.0" />
    <PackageReference Include="Microsoft.Extensions.Options" Version="6.0.0" />
  </ItemGroup>

  <ItemGroup Condition=" '$(TargetFramework)' == 'net8.0' ">
    <PackageReference Include="Microsoft.Extensions.Hosting.Abstractions" Version="8.0.0" />
    <PackageReference Include="Microsoft.Extensions.Logging.Abstractions" Version="8.0.0" />
    <PackageReference Include="Microsoft.Extensions.Options" Version="8.0.0" />
  </ItemGroup>
	
</Project>
```
**IConsumerServiceSelector.Default.cs**

```
#if NET8_0_OR_GREATER

#else

#endif

```

**Comment by yang-xiaodong** [comment]
Fixed in version 8.0.0
